### DISCUSSION TITLE
Issue with `ls` but not sure if intended

### DISCUSSION THREAD

**Discussion by ykh-1992** [git-for-windows/git#3682] [Q&A] [answered]
Hello,

I found an issue with `ls` but I'm not sure if it's meant to work this way.
The issue I find is when working with directories under OneDrive. Whenever I try to list files under such directories using `ls`, it goes ahead and downloads the entire directory I'm in before displaying them.
As far as I understand, `ls` should only list things, not download them. Is this behavior intended? Or is it incorrect?

Appreciate any help.

**Comment by ykh-1992** [comment]
Attached is screenshot of Windows telling me that `ls` is downloading files.

![image](https://user-images.githubusercontent.com/18620109/152303610-49c9d5d2-3729-47fb-b1ba-880eb82fdce6.png)

**Comment by rimrul** [comment, marked as answer]
> As far as I understand, `ls` should only list things, not download them. 

Well, it's more complicated that that. `ls` is supposed to show you information about files. At its most basic, that's the names of the files in the current directory. And it needs to gather this information before it can display it. Since `ls` is designed for unix systems it uses `stat()` (or similar) for that. `stat()` provides a bunch of info about a file, including its size and the time it was last modified.

But Windows doesn't have a direct `stat()` equivalent, so the MSYS2 runtime has to provide a `stat()` implementation that cobbles together the required information from various Windows functions. One or more of them require the actual file to be present, so OneDrive downloads the file.

> Is this behavior intended? Or is it incorrect?

I'd argue it's neither. It's not intended by the `ls` developers, since to them a file is a local file and not something that gets downloaded on demand when you query info about it, but to one drive these requests for information are a programm trying to open a file, so it should be downloaded. It's an unfortunate side effect that is to be expected based on how these two applications work.

**Reply by ykh-1992** [reply]
So in this case, is there an alternative to listing files in a directory without invoking OneDrive to downloading them? Otherwise, is it possible to modify `ls` behavior such that it doesn't download files?
I'm a complete newbie to all this, so any pointers would help immensely.

**Reply by dscho** [reply]
I guess a better way would be to use `cmd /c dir` in such a case. But note! If there is _any_ shell script in there, the `ls -l` will show it executable, but the `dir` won't (and yes, this is why the files are inspected, you have to open them to find out whether they're executable scripts or not).
